research/job_referencing/index.md

```diff
@@ -0,0 +1,217 @@
+authors:
+  - mmassari
+sidebar_position: 1
+---
+
+# Job referencing
+
+## Already implemented use cases
+
+### COPR build depends on SRMP build
+
+[Here](https://github.com/packit/packit-service/blob/16236087fb8071c83b3802464f3af504f4fb933e/packit_service/worker/handlers/copr.py#L136-L203) we build both the SRPM and the COPR packages (and the COPR package depends on the SRPM one).
+We don't need any special data to be saved in the database, we just check for the existance of the SRPM build.
+This code is triggered by a `CoprBuildStartEvent` event, this event is sent by COPR both for the SRPM package build and for the COPR package build, in this way we run twice the code and we can deal with the dependency.
+
+### Tests depends on COPR builds
+
+[Here](https://github.com/packit/packit-service/blob/16236087fb8071c83b3802464f3af504f4fb933e/packit_service/worker/handlers/testing_farm.py#L202-L241) we check that a COPR build exist for every target we should run a test for.
+If not build is found we run a `CoprBuldHandler` instance.
+For every `CoprBuildEndEvent` the `CoprBuildEndHandler` is able to run again the `TestingFarmHandler` if needed.
+No special data is stored in the database, the COPR build is searched using: project name, commit sha, copr build owner and target.
+
+## New use cases
+
+### Tests depends on specified COPR builds
+
+Since it is possible to add identifiers for jobs, we allow defining multiple Copr build jobs and multiple TF jobs. In that case, Packit Service doesn't know what Copr build job to use for the particular TF job.
+
+For this purpose, we could introduce a new field for job configs which would allow the referencing, for example:
+
+```yaml
+- job: copr_build
+identifier: "build1"
+trigger: pull_request
+actions: ...
+- job: copr_build
+identifier: "build2"
+trigger: pull_request
+actions: ...
+- job: tests
+identifier: "tests2"
+trigger: pull_request
+build_identifier: "build1"
+- job: tests
+identifier: "tests2"
+trigger: pull_request
+build_identifier: "build2"
+```
+
+#### The easiest solution I see
+
+I don't think we need anything more in the database to be able to retrieve the copr build id. We need the data given in the `copr_build` job matching the specified `build_identifier` name in the `tests` job, plus all the data we are already using in the search like the `commit_sha`.
+We need to add the `build_identifier` key in config.
+We need to modify [this code](https://github.com/packit/packit-service/blob/16236087fb8071c83b3802464f3af504f4fb933e/packit_service/worker/handlers/testing_farm.py#L202-L241) in a way that it is able to retrieve the proper `build_id`.
+
+#### `build_identifier` key name or?
+
+Another suggestion was to use `after` as a name for the key. I like it but our implementation, if made as specified above, will not be generic. For this reason I would rather prefer `after_build` or `build_identifier` to mark the fact that in the test job we are waiting for a build job.
```

Comment:
This to me means something like being able to start from any handler any other handler in a generic way. I don't like the idea at the moment, maybe after a refactoring and using new tools this can become easier but at the moment I think this is risky.

Reply:
Yes, ideally, use some framework for that. I am worried resolving all the dependencies one by one might increase the code complexity a lot compared to a generic solution.

Reply:
agreed, the code handling the current dependencies (e.g. build-> TF) is already making the code complexity much worse